### ISSUE TITLE
websocket disconnect  every minute and raise 503 error.

### ISSUE THREAD

**Issue by xkungfu** [sanic-org/sanic#2078]
**Describe the bug**
from client, every 25 seconds send a "ping",
on server side, when received "ping", send a "pong".

on client side, after sent a "ping", wait 15 seconds, if not receive "pong", then reconnect.

websocket disconnect  about every minute.

**Code snippet**
server.py:
```python
from sanic import Sanic
app = Sanic("websocket_example")
@app.websocket('/wsserver')
async def feed(request, ws):
    while True:
        data = 'pong'
        print('Sending: ' + data)
        await ws.send(data)
        data = await ws.recv()
        print('Received: ' + data)
if __name__ == "__main__":
    app.run(host="0.0.0.0", port=8001)

```

**Expected behavior**
keep connection!

**Environment (please complete the following information):**
 - OS: ubuntu20.04
 - Version : sanic 21.3.1

**Additional context**
on server side log:
```
[2021-03-22 06:28:15 +0800] [3380798] [INFO] Goin' Fast @ http://0.0.0.0:8001
[2021-03-22 06:28:15 +0800] [3380798] [INFO] Starting worker [3380798]
Sending: pong
Received: ping
Sending: pong
Received: ping
Sending: pong
Received: ping
Sending: pong
[2021-03-22 06:29:35 +0800] - (sanic.access)[INFO][42.185.74.69:34866]: GET ws://42.3.27.246/wsserver  503 -1
Sending: pong
Received: ping
Sending: pong
Received: ping
Sending: pong
Received: ping
Sending: pong
[2021-03-22 06:30:52 +0800] - (sanic.access)[INFO][42.185.74.69:34911]: GET ws://42.3.27.246/wsserver  503 -1
Sending: pong
Received: ping
Sending: pong
Received: ping
Sending: pong
Received: ping
Sending: pong
[2021-03-22 06:32:09 +0800] - (sanic.access)[INFO][42.185.74.69:35021]: GET ws://42.3.27.246/wsserver  503 -1

```

**Comment by xkungfu** [comment]
I have test some different codes, the same error on them.

I have no idea now, please check if it is a bug or my code is wrong?

anyone faced the same issue?

**Comment by ahopkins** [comment]
It is tough to say without seeing your client code as well. One of the first thing that jumps out at me is that you have both the consumer (`ws.recv()`) and the producer (`ws.send()`) running consecutively. Therefore, each of those operations will block the other from happening. The handler will pause at each `await` until it can be resolved. Depending upon your client logic, that might cause the described behavior.

**Comment by xkungfu** [comment]
thanks for reply.

**Comment by xkungfu** [comment]
thanks for reply.

I am almost certain it is a bug in sanic new version  21.3.0 and 21.3.1.

on sanic 21.12.0-21.12.3, the codes work smoothly.

please have a deep check.

**Comment by ahopkins** [comment]
Can you provide your client code?

**Comment by xkungfu** [comment]
the client very simple: 
1\every 25 seconds send a "ping",
2\after sent a "ping", wait 15 seconds, if not receive "pong", then reconnect.

the  (ws.recv()) and  (ws.send()) running consecutively, but it is copy from your official documents.

now I add a new logger about reconnection to show the bug clearly. 

`
@app.websocket("/wsserver")
async def feed(request, ws):
    try:

        logger.debug(f"the client re-connected!")

        while True:

            #print("request:　", request)
            #print("ws:　", ws)

            name = await ws.recv()
            print(f"< {name}")

            greeting = f"pong"

            await ws.send(greeting)
            print(f"> {greeting}")

    except websockets.exceptions.ConnectionClosedOK:
        logger.debug("/wsserver ConnectionClosedOK")
    except websockets.exceptions.ConnectionClosedError:
        logger.debug("/wsserver ConnectionClosedError")
    except sanic.websocket.ConnectionClosed:
        logger.debug("/wsserver ConnectionClosed: " + str(e))
    except websockets.exceptions.WebSocketException as e:
        logger.debug("/wsserver WebSocketException: " + str(e))
    except concurrent.futures.CancelledError:
        logger.debug("/wsserver CancelledError")
    except BaseException as e:
        logger.debug("/wsserver BaseException: " + str(type(e)))
`


```

[2021-03-22 13:07:35 +0800] [3463351] [INFO] Goin' Fast @ http://0.0.0.0:8001
[2021-03-22 13:07:35 +0800] [3463351] [INFO] Starting worker [3463351]
[2021-03-22 13:07:40 +0800] [3463351] [DEBUG] the client re-connected!
< ping
> pong
< ping
> pong
< ping
> pong
[2021-03-22 13:08:42 +0800] [3463351] [DEBUG] /wsserver BaseException: <class 'asyncio.exceptions.CancelledError'>
[2021-03-22 13:08:43 +0800] - (sanic.access)[INFO][104.233.191.163:34936]: GET ws://42.3.27.246/wsserver  503 -1
[2021-03-22 13:08:43 +0800] [3463351] [ERROR] Connection lost before response written @ ('104.233.191.163', 34936) <Request: GET /wsserver>
[2021-03-22 13:08:54 +0800] [3463351] [DEBUG] the client re-connected!
[2021-03-22 13:09:05 +0800] [3463351] [DEBUG] /wsserver BaseException: <class 'asyncio.exceptions.CancelledError'>
[2021-03-22 13:09:05 +0800] [3463351] [ERROR] Connection lost before response written @ ('104.233.191.163', 35081) <Request: GET /wsserver>
[2021-03-22 13:09:06 +0800] [3463351] [DEBUG] the client re-connected!
[2021-03-22 13:09:17 +0800] [3463351] [DEBUG] /wsserver BaseException: <class 'asyncio.exceptions.CancelledError'>
[2021-03-22 13:09:17 +0800] [3463351] [ERROR] Connection lost before response written @ ('104.233.191.163', 35100) <Request: GET /wsserver>
[2021-03-22 13:09:17 +0800] [3463351] [DEBUG] the client re-connected!
[2021-03-22 13:09:28 +0800] [3463351] [DEBUG] /wsserver BaseException: <class 'asyncio.exceptions.CancelledError'>
[2021-03-22 13:09:28 +0800] [3463351] [ERROR] Connection lost before response written @ ('104.233.191.163', 35107) <Request: GET /wsserver>
```

**Comment by xkungfu** [comment]
the client written in C++ : 

```
	countforping = 0;
	countforpong = 0;
	waitingpong = false;

while true:

.....
        if (countforping > 50) {
	        std::string ping = "ping";
	        sesh->send_message(ping);
	        countforping = 0;
	        waitingpong = true;
        }
        
        if (waitingpong == true) countforpong++;
        
        if (msgfromserver == "pong") {
		        cout << "msgfromserver==pong ... ... ... ... ... ... ... ... ... ... ... ... ... ... ... ... " << endl;
		        waitingpong = false;
		        countforpong = 0;
	        }
        
        if(countforpong > 30){
	        cout << "countforpong > DISCONNECTED  need reconnect ... " << countforpong << endl;
	        countforpong = 0;
        
        -----------------------------------------------------
        --> excute reconnection logic
        -----------------------------------------------------
        
        }
......

```

**Comment by ahopkins** [comment]
`countforping` is never incremented. I think you left out something. 

Regardless, I did find the issue and can confirm this is a bug in that the response timeout is being triggered. It should be ignored here.

**Comment by ashleysommer** [comment]
@ahopkins 
Is this a bug in the new streaming server in the response handling? Is the response timeout triggering on the websocket when it should not? Should the response timeout simply not apply to websocket connections?

**Comment by ashleysommer** [comment]
@xkungfu can you try to set the sanic "RESPONSE_TIMEOUT" config to something very large (eg, 99999) to see if that is a viable workaround to your problem?

**Comment by ahopkins** [comment]
@ashleysommer 

> Is this a bug in the new streaming server in the response handling? 

Streaming server. See #2081 for the fix and I think you will clearly see the issue.

> Is the response timeout triggering on the websocket when it should not? Should the response timeout simply not apply to websocket connections?

It probably should not. I am still not 100% thrilled with the websockets integration, and I think this is something that should be addressed in #2000. However, completely disabling it brings up the potential of a non-responsive handler. Ultimately, I think a better ping/pong mechanism should be implemented out of the box for the dev without needing to do anything. Even if a response is not being sent to the client, there seems like there should be something to monitor the handler. Perhaps there is some sort of a timeout loop on `ws.recv()` under the hood to report back to the Sanic protocol that its still waiting.

> @xkungfu can you try to set the sanic "RESPONSE_TIMEOUT" config to something very large (eg, 99999 seconds) to see if that is a viable workaround to your problem?

This would solve the problem. But, then you have the issue of no response timeout protection on any other responses.

Which then brings up another question that has been raised before. I will need to go digging to find the discussion, and will open a new issue. That is having a temporary config override:

```python
with app.config.override(RESPONSE_TIMEOUT=99999):
    do_something()

# or

app.route("/", config_override={"RESPONSE_TIMEOUT":99999 })(...)
```

**Comment by xkungfu** [comment]
> `countforping` is never incremented. I think you left out something.
> 
> Regardless, I did find the issue and can confirm this is a bug in that the response timeout is being triggered. It should be ignored here.

sorry, my mistake missing  "	countforping++; ".  the codes was edited and updated.

**Comment by xkungfu** [comment]
> @ashleysommer
> 
> > Is this a bug in the new streaming server in the response handling?
> 
> Streaming server. See #2081 for the fix and I think you will clearly see the issue.
> 
> > Is the response timeout triggering on the websocket when it should not? Should the response timeout simply not apply to websocket connections?
> 
> It probably should not. I am still not 100% thrilled with the websockets integration, and I think this is something that should be addressed in #2000. However, completely disabling it brings up the potential of a non-responsive handler. Ultimately, I think a better ping/pong mechanism should be implemented out of the box for the dev without needing to do anything. Even if a response is not being sent to the client, there seems like there should be something to monitor the handler. Perhaps there is some sort of a timeout loop on `ws.recv()` under the hood to report back to the Sanic protocol that its still waiting.
> 
> > @xkungfu can you try to set the sanic "RESPONSE_TIMEOUT" config to something very large (eg, 99999 seconds) to see if that is a viable workaround to your problem?
> 
> This would solve the problem. But, then you have the issue of no response timeout protection on any other responses.
> 
> Which then brings up another question that has been raised before. I will need to go digging to find the discussion, and will open a new issue. That is having a temporary config override:
> 
> ```python
> with app.config.override(RESPONSE_TIMEOUT=99999):
>     do_something()
> 
> # or
> 
> app.route("/", config_override={"RESPONSE_TIMEOUT":99999 })(...)
> ```

I would use version 20.12.3 now because no error on this version, and waiting for next version.

I think keeping live for websocket is very seriously and difficult problem,  it seems that there is not any simple solution. this logic about "ping and pong and reconnect when lost“  is my test. and I think it is not a good solution. but I don't know others.

thank you. good luck!

**Comment by ahopkins** [comment]
@xkungfu See #2081
